File: src/bun.js/api/Timer/ImmediateObject.zig
const Self = @This();

const RefCount = bun.ptr.RefCount(@This(), "ref_count", deinit, .{});
pub const ref = RefCount.ref;
pub const deref = RefCount.deref;

pub const js = jsc.Codegen.JSImmediate;
pub const toJS = js.toJS;
pub const fromJS = js.fromJS;
pub const fromJSDirect = js.fromJSDirect;

ref_count: RefCount,
event_loop_timer: EventLoopTimer = .{
    .next = .epoch,
    .tag = .ImmediateObject,
},
internals: TimerObjectInternals,

pub fn init(
    globalThis: *JSGlobalObject,
    id: i32,
    callback: JSValue,
    arguments: JSValue,
) JSValue {
    // internals are initialized by init()
    const immediate = bun.new(Self, .{ .ref_count = .init(), .internals = undefined });
    const js_value = immediate.toJS(globalThis);
    defer js_value.ensureStillAlive();
    immediate.internals.init(
        js_value,
        globalThis,
        id,
        .setImmediate,
        0,
        callback,
        arguments,
    );

    if (globalThis.bunVM().isInspectorEnabled()) {
        Debugger.didScheduleAsyncCall(
            globalThis,
            .DOMTimer,
            ID.asyncID(.{ .id = id, .kind = .setImmediate }),
            true,
        );
    }

    return js_value;
}

fn deinit(self: *Self) void {
    self.internals.deinit();
    bun.destroy(self);
}

pub fn constructor(globalObject: *jsc.JSGlobalObject, callFrame: *jsc.CallFrame) !*Self {
    _ = callFrame;
    return globalObject.throw("Immediate is not constructible", .{});
}

/// returns true if an exception was thrown
pub fn runImmediateTask(self: *Self, vm: *VirtualMachine) bool {
    return self.internals.runImmediateTask(vm);
}

pub fn toPrimitive(self: *Self, _: *jsc.JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.toPrimitive();
}

pub fn doRef(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.doRef(globalThis, callFrame.this());
}

pub fn doUnref(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.doUnref(globalThis, callFrame.this());
}

pub fn hasRef(self: *Self, _: *JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.hasRef();
}

pub fn finalize(self: *Self) void {
    self.internals.finalize();
}

pub fn getDestroyed(self: *Self, globalThis: *JSGlobalObject) JSValue {
    _ = globalThis;
    return .jsBoolean(self.internals.getDestroyed());
}

pub fn dispose(self: *Self, globalThis: *JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    self.internals.cancel(globalThis.bunVM());
    return .js_undefined;
}

const Debugger = @import("../../Debugger.zig");
const bun = @import("bun");

const EventLoopTimer = bun.api.Timer.EventLoopTimer;
const ID = bun.api.Timer.ID;
const TimerObjectInternals = bun.api.Timer.TimerObjectInternals;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;
const VirtualMachine = jsc.VirtualMachine;

File: src/runtime/timer/ImmediateObject.rs
use core::cell::Cell;

use bun_jsc::virtual_machine::VirtualMachine;
use bun_jsc::{JSGlobalObject, JSValue};
use bun_ptr::{JsCell, RefPtr, ThisPtr};

use super::{EventLoopTimer, IdMap, Kind, Maps, TimerObject, TimerObjectInternals};

// `jsc.Codegen.JSImmediate` — the C++ JSCell wrapper stays generated; this
// struct is the `m_ctx` payload. Struct + `RefCounted`/`Drop` impls + the
// forwarder host-fns (`to_primitive`/`do_ref`/`do_unref`/`has_ref`/
// `get_destroyed`/`dispose`/`constructor`/`finalize`/`init_with`) — see
// `impl_timer_object!` in `super` (timer/mod.rs).
super::impl_timer_object!(ImmediateObject, ImmediateObject, "Immediate");

impl TimerObject for ImmediateObject {
    #[inline]
    fn internals(&self) -> &TimerObjectInternals {
        &self.internals
    }
    #[inline]
    fn event_loop_timer(&self) -> &JsCell<EventLoopTimer> {
        &self.event_loop_timer
    }
    #[inline]
    fn heap_ref(&self) -> &Cell<Option<RefPtr<Self>>> {
        &self.heap_ref
    }
    #[inline]
    fn id_map(maps: &mut Maps, _kind: Kind) -> &mut IdMap<Self> {
        &mut maps.set_immediate
    }
}

impl ImmediateObject {
    pub(crate) fn init(
        global: &JSGlobalObject,
        id: i32,
        callback: JSValue,
        arguments: JSValue,
    ) -> JSValue {
        Self::init_with(global, id, Kind::SetImmediate, 0, callback, arguments)
    }

    /// Reached from `bun_jsc::event_loop` via `__bun_run_immediate_task`
    /// (definer in [`crate::dispatch`]). Returns `true` if an exception was
    /// thrown. `this` carries the immediate queue's ref; it may be gone once
    /// this returns.
    #[inline]
    pub(crate) fn run_immediate_task(this: ThisPtr<Self>, vm: &VirtualMachine) -> bool {
        TimerObject::run_immediate_task(this, vm)
    }

    /// Release the immediate queue's ref without running the callback (VM
    /// teardown). `this` may be gone once this returns.
    #[inline]
    pub(crate) fn cancel_pending(this: ThisPtr<Self>, _vm: &VirtualMachine) {
        TimerObject::cancel_pending_immediate(this);
    }
}
